File: sdk/python/captcha-font-sdk/captcha_font_sdk/group_id.py
from __future__ import annotations

from typing import List, Tuple

RGB = Tuple[int, int, int]


def rgb_at(pix: List[Tuple[int, ...]], width: int, x: int, y: int) -> RGB:
    p = pix[y * width + x]
    return int(p[0]), int(p[1]), int(p[2])


def group_id_from_pixels(pix: List[Tuple[int, ...]], width: int, height: int) -> str:
    corners = [
        rgb_at(pix, width, 0, 0),  # lt
        rgb_at(pix, width, width - 1, 0),  # rt
        rgb_at(pix, width, 0, height - 1),  # lb
        rgb_at(pix, width, width - 1, height - 1),  # rb
    ]
    return f"{width}x{height}|" + "|".join(f"{r},{g},{b}" for r, g, b in corners)

File: typescript-sdk/src/group-id.ts
import { RGBA } from './types';

export type RGB = [number, number, number];

function rgbAt(pixels: RGBA[], width: number, x: number, y: number): RGB {
  const p = pixels[y * width + x];
  return [p[0], p[1], p[2]];
}

export function groupIdFromPixels(pixels: RGBA[], width: number, height: number): string {
  const corners: RGB[] = [
    rgbAt(pixels, width, 0, 0),
    rgbAt(pixels, width, width - 1, 0),
    rgbAt(pixels, width, 0, height - 1),
    rgbAt(pixels, width, width - 1, height - 1)
  ];
  return `${width}x${height}|` + corners.map(([r, g, b]) => `${r},${g},${b}`).join('|');
}
